content/apply.js

```diff
@@ -1,18 +1,18 @@
-/* eslint no-var: 0 */
 /* global msg API prefs createStyleInjector */
-/* exported APPLY */
 'use strict';
 
-// some weird bug in new Chrome: the content script gets injected multiple times
-// define a constant so it throws when redefined
-const APPLY = (() => {
-  const CHROME = chrome.app ? parseInt(navigator.userAgent.match(/Chrom\w+\/(?:\d+\.){2}(\d+)|$/)[1]) : NaN;
+// Chrome reruns content script when documentElement is replaced.
+// Note, we're checking against a literal `1`, not just `if (truthy)`,
+// because <html id="INJECTED"> is exposed per HTML spec as a global variable and `window.INJECTED`.
+
+// eslint-disable-next-line no-unused-expressions
+self.INJECTED !== 1 && (() => {
```

Comment:
> should not trigger any errors.

Why? Can you post a repro?

Reply:
It's a known thing so I'd rather not spend my time on that.

Reply:
Eh, couldn't get it out of my mind, here's the chromium's source code reference: [link](https://source.chromium.org/chromium/chromium/src/+/master:extensions/renderer/script_injection_manager.cc;l=142;drc=a7af154233103371fc11baf4e0ede4d1dba5f939?sq=package:chromium).

Reply:
> there's no benefit in showing these errors

They notify developers that something went wrong. [Here is an example](https://github.com/openstyles/stylus/pull/843#issuecomment-584110590). Without the error, we will never know that there is a content script reinjection bug when the browser starts.

> here's the chromium's source code reference: [link](https://source.chromium.org/chromium/chromium/src/+/master:extensions/renderer/script_injection_manager.cc;l=142;drc=a7af154233103371fc11baf4e0ede4d1dba5f939?sq=package:chromium).

I don't understand Chrome's structure. Could you describe some steps? I can help to file a bug on crbug.com if you don't have time.

Reply:
What reinjection bug? The only negative effect I see is those useless errors. Anyway I can't comment on that as this can be something your rewrite has broken and I never analyzed the new code. So I still see exactly zero benefit in these errors.

I also won't help you with crbug as I don't really see it as a bug but rather an architectural peculiarity, moreover I have 99% certainty it won't get fixed in the next 10 years. If you didn't know there are hundreds of extensions-related abandoned and ignored reports on crbug which accumulated over the last 10 years, especially the last 5 when only one developer was left maintaining the API internals and fixing the most critical bugs, the rest of the extensions team has moved onto other projects or components and only occasionally fix a bug or two.

Is this really something worth all this fuss?

Reply:
> What reinjection bug?

The bug mentioned by @narcolepticinsomniac is not caused by this PR. It can be inconsistently produced in master. I think the problem is that we use message ping to detect if content scripts are available:
https://github.com/openstyles/stylus/blob/438fdebc5c7ebd2e7f728a7295f6db754d575b6a/background/content-scripts.js#L34

When the browser starts, there is a chance that the pong callback/content script is not available yet but the background script started pinging. As a result, content scripts defined in manifest are injected and background also tries to inject content scripts again. Therefore the error occurs.

I guess this can be produced consistently if there is a heavy processing content script (which can be injected by other extensions) is injected before `apply.js`.

> this alleged bug is fixed.

IMHO, ignoring the error doesn't mean the bug is fixed.

> Is this really something worth all this fuss?

As a developer, I hope we can leave some reference/indication/workaround for who hit the same bug.

Reply:
>As a developer, I hope we can leave some reference/indication/workaround for who hit the same bug.

Leave a comment then. If these errors are caused by a browser bug, or they just don't matter, I'd appreciate not having to constantly see them anymore.

Reply:
People happily lived with this bug present for years and no one died. Once this reinjection bug is fixed there'll be no benefit in showing these errors. So why are we still discussing this boring topic?

Reply:
Bugfix in #849.

Reply:
> Once this reinjection bug is fixed there'll be no benefit in showing these errors.

IMHO, fixing one bug doesn't mean that there are no other bugs, or there will be no bugs introduced by us or the browser. So it is still worth keeping the `const`.